### PR TITLE
refactor(config): refactor the config object.

### PR DESCRIPTION
**Pull request by hansl**
This also will load `angular-cli.json` in the HOME directory as a fallback, supports more stuff from the JSON Schema (like default values) than the old one, and actually verify that what you inputs is the right thing.

This will be its own NPM package at some point, as other people will probably be interested in having a JSON Schema loader that gives type safety and provides fallbacks and metadata.

Closes https://github.com/angular/angular-cli/issues/1763

### CODE REVIEW

File: addon/ng2/commands/set.ts [hunk 1]
```diff
@@ -11,10 +12,38 @@ const SetCommand = Command.extend({
...
+    return +raw;
+  },
+
   run: function (commandOptions, rawArgs): Promise<void> {
     return new Promise(resolve => {
-      const config = new CliConfig();
-      config.set(rawArgs[0], rawArgs[1], commandOptions.force);
+      const [jsonPath, rawValue] = rawArgs;
+      const config = CliConfig.fromProject();
+      const type = config.typeOf(rawArgs[0]);
```
Comment (Brocco) [inline addon/ng2/commands/set.ts]: Can we use `config.typeOf(jsonPath);` here since it's already destructured?

File: addon/ng2/commands/set.ts [hunk 2]
```diff
@@ -11,10 +12,38 @@ const SetCommand = Command.extend({
...
+      throw new SilentError(`Invalid number value: "${raw}"`);
+    }
+    return +raw;
+  },
+
   run: function (commandOptions, rawArgs): Promise<void> {
     return new Promise(resolve => {
-      const config = new CliConfig();
-      config.set(rawArgs[0], rawArgs[1], commandOptions.force);
+      const [jsonPath, rawValue] = rawArgs;
```
Comment (filipesilva) [inline addon/ng2/commands/set.ts]: `rawValue` should just be `value`, no need for `let value = rawValue;` in line 36.

File: addon/ng2/commands/set.ts [hunk 3]
```diff
@@ -11,10 +12,38 @@ const SetCommand = Command.extend({
...
+      const config = CliConfig.fromProject();
+      const type = config.typeOf(rawArgs[0]);
+      let value = rawValue;
+
+      switch (type) {
+        case 'boolean': value = this.asBoolean(rawValue); break;
+        case 'number': value = this.asNumber(rawValue); break;
+        case 'string': value = rawValue; break;
+
```
Comment (filipesilva) [inline addon/ng2/commands/set.ts]: Is this newline intended?

File: addon/ng2/commands/test.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ module.exports = TestCommand.extend({
   ],
 
   run: function (commandOptions) {
-    this.project.ngConfig = this.project.ngConfig || config.CliConfig.fromProject();
+    this.project.ngConfig = this.project.ngConfig || CliConfig.fromProject();
```
Comment (filipesilva) [inline addon/ng2/commands/test.ts]: Should be `|| CliConfig.fromProject().config;` I think.

File: addon/ng2/commands/set.ts (hunk 1 above)
Comment (hansl) [inline addon/ng2/commands/set.ts]: done.

File: addon/ng2/commands/set.ts (hunk 2 above)
Comment (hansl) [inline addon/ng2/commands/set.ts]: it's const, moved it to let.

File: addon/ng2/commands/set.ts (hunk 3 above)
Comment (hansl) [inline addon/ng2/commands/set.ts]: Basically between options and default. I can remove if you like.

File: addon/ng2/models/webpack-config.ts [hunk 5]
```diff
@@ -21,15 +21,16 @@ export class NgCliWebpackConfig {
...
 
-    appConfig.outDir = outputDir || appConfig.outDir; 
+    appConfig.outDir = outputDir || appConfig.outDir;
 
     this.webpackBaseConfig = getWebpackCommonConfig(this.ngCliProject.root, environment, appConfig);
     this.webpackDevConfigPartial = getWebpackDevConfigPartial(this.ngCliProject.root, appConfig);
     this.webpackProdConfigPartial = getWebpackProdConfigPartial(this.ngCliProject.root, appConfig);
 
-    if (CliConfig.fromProject().apps[0].mobile){
+    if (CliConfig.fromProject().config.apps[0].mobile){
```
Comment (filipesilva) [inline addon/ng2/models/webpack-config.ts]: Better to do `if (appConfig.mobile){`

File: addon/ng2/commands/test.ts (hunk 4 above)
Comment (hansl) [inline addon/ng2/commands/test.ts]: Yeah I should do something about that :/  done.

Comment (hansl) [inline addon/ng2/commands/test.ts]: Actually this is correct. We use `ngConfig.config` everywhere where it's needed. I do think `config` is a misnomer and might be better off as `root`, but I have to refactor some usage of this anyway. That might happen later on (with Feenix).

File: addon/ng2/commands/set.ts (hunk 3 above)
Comment (filipesilva) [inline addon/ng2/commands/set.ts]: No need, just wanted to be sure it was intended.

File: addon/ng2/commands/test.ts (hunk 4 above)
Comment (filipesilva) [inline addon/ng2/commands/test.ts]: But won't it break other commands if they are run in sequence? Everything else expects `this.project.ngConfig` to have `config` instead. 

Or the other way around, if nothing else breaks by having one command assign the wrong thing to `this.project.ngConfig`, maybe this is unnecessary.

Comment (filipesilva) [inline addon/ng2/commands/test.ts]: Weird, I thought I saw this being attributed loads of times.. but it's only in `index.js` now that I double check. My bad.

File: package.json [hunk 6]
```diff
@@ -105,9 +106,11 @@
     ]
   },
   "devDependencies": {
+    "@types/node": "^6.0.36",
```
Comment (filipesilva) [inline package.json]: Is this doing anything? In the generated project I had to specify `typeRoots` for it to work from a subdir.

File: package.json [hunk 7]
```diff
@@ -118,6 +121,7 @@
     "sinon": "^1.17.3",
     "through": "^2.3.8",
     "tree-kill": "^1.0.0",
+    "ts-node": "^1.2.3",
```
Comment (filipesilva) [inline package.json]: What's `ts-node` for? I can't find it being used in this PRs changes.

File: addon/ng2/models/webpack-config.ts (hunk 5 above)
Comment (hansl) [inline addon/ng2/models/webpack-config.ts]: Done.

File: package.json (hunk 7 above)
Comment (hansl) [inline package.json]: Removed. Probably from a previous version of this PR.

File: package.json (hunk 6 above)
Comment (hansl) [inline package.json]: yep. It works.

Comment (hansl) [inline package.json]: Note that because we're not using a real strong compiler, typings are kind of ignored. I want to fix that at some point soon.

**Comment by filipesilva** [comment]
What is the `addon/ng2/models/json-schema/schema-tree.ts` empty file?

**Comment by filipesilva** [comment]
Can you also add some information on the README about how to use `ng config`? We have none atm and mostly ask users to edit the file directly.

**Comment by hansl** [comment]
Re empty file, contains stuff now.
there's no change from the CLI. `ng set` and `ng get` are still the same.

**Comment by filipesilva** [comment]
It's true that they are the same, but we have no documentation whatsoever for users to use them, so no one but us knows how to use this functionality.

**Pull request merged**